RxSwift/Observables/Implementations/Materialize.swift

```diff
@@ -0,0 +1,72 @@
+//
+//  Materialize.swift
+//  Rx
+//
+//  Created by sergdort on 08/03/2017.
+//  Copyright © 2017 Krunoslav Zaher. All rights reserved.
+//
+
+import Foundation
+
+public protocol EventType {
+    associatedtype E
+    func asEvent() -> Event<E>
+}
+
+extension Event : EventType {
+    public typealias E = Element
+    public func asEvent() -> Event<E> { return self }
+}
+
+fileprivate final class MaterializeSink<Element>: Disposable, ObserverType {
```

Comment:
Not sure I understand why we can't use it. As far as I can tell we can. `Event<Element>` is no different than any other value.

```
fileprivate final class MaterializeSink<Element, O: ObserverType>
    : Sink<O>
    , ObserverType where O.E == Event<Element> {
    typealias E = Element
    override init(observer: O, cancel: Cancelable) {
        super.init(observer: observer, cancel: cancel)
    }

    func on(_ event: Event<Element>) {
        self.forwardOn(.next(event))
        if event.isStopEvent {
            self.dispose()
        }
    }
}

internal class Materialize<Element>: Producer<Event<Element>> {
    let source: Observable<Element>

    init(source: Observable<Element>) {
        self.source = source
    }

    override func run<O : ObserverType >(_ observer: O, cancel: Cancelable) -> (sink: Disposable, subscription: Disposable) where O.E == Event<Element> {
        let sink = MaterializeSink(observer: observer, cancel: cancel)
        let subscription = source.subscribe(sink)

        return (sink: sink, subscription: subscription)
    }
}
```